python/src/iceberg/catalog/base.py

```diff
@@ -0,0 +1,247 @@
+#  Licensed to the Apache Software Foundation (ASF) under one
+#  or more contributor license agreements.  See the NOTICE file
+#  distributed with this work for additional information
+#  regarding copyright ownership.  The ASF licenses this file
+#  to you under the Apache License, Version 2.0 (the
+#  "License"); you may not use this file except in compliance
+#  with the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing,
+#  software distributed under the License is distributed on an
+#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+#  KIND, either express or implied.  See the License for the
+#  specific language governing permissions and limitations
+#  under the License.
+
+from abc import ABC, abstractmethod
+from typing import Dict, List, Optional, Set, Tuple, Union
+
+from iceberg.schema import Schema
+from iceberg.table.base import PartitionSpec, Table
+
+Identifier = Tuple[str, ...]
+Properties = Dict[str, str]
+
+
+class Catalog(ABC):
+    """Base Catalog for table operations like - create, drop, load, list and others.
+
+    The catalog table APIs accept a table identifier, which is fully classified table name. The identifier can be a string or
+    tuple of strings. If the identifier is a string, it is split into a tuple on '.'. If it is a tuple, it is used as-is.
+
+    The catalog namespace APIs follow a similar convention wherein they also accept a namespace identifier that can be a string
+    or tuple of strings.
+
+    Attributes:
+        name(str): Name of the catalog
+        properties(Properties): Catalog properties
+    """
+
+    def __init__(self, name: str, properties: Properties):
```

Comment:
I'll continue my quest for data classes here as well. Instead of creating the getters/setters, we could also let the dataclasses do this:
```python
@dataclass
class Catalog(ABC):

    name: str =  field()
    name: Properties =  field(default_factory=dict)
```
This will automatically generate the getters and setters.

Reply:
If we want to go full blown pythonic in the future, we could also validate these using Pydantic validators: https://pydantic-docs.helpmanual.io/usage/validators/

Reply:
I think using `@dataclass` is a good idea, but we can move on here and refactor it later.